test/videojs.ads.test.js

```diff
@@ -1,1025 +1,732 @@
+  this.player.trigger('playing');
+  assert.strictEqual(this.player.ads.state, 'content-playback');
+  this.player.ads.startLinearAdMode();
+  assert.strictEqual(spy.callCount, 2, 'adstart should have fired');
+});
+
+QUnit.test('calling endLinearAdMode() in any state but ad-playback does not trigger adend', function(assert) {
+  assert.expect(13);
+  var spy = sinon.spy();
+  this.player.on('adend', spy);
+  assert.strictEqual(this.player.ads.state, 'content-set');
+  this.player.ads.endLinearAdMode();
+  assert.strictEqual(spy.callCount, 0, 'adend should not have fired');
+  this.player.trigger('adsready');
+  assert.strictEqual(this.player.ads.state, 'ads-ready');
+  this.player.ads.endLinearAdMode();
+  assert.strictEqual(spy.callCount, 0, 'adend should not have fired');
+  this.player.tech.trigger('play');
+  assert.strictEqual(this.player.ads.state, 'preroll?');
+  this.player.ads.endLinearAdMode();
+  assert.strictEqual(spy.callCount, 0, 'adend should not have fired');
+  this.player.trigger('adtimeout');
+  assert.strictEqual(this.player.ads.state, 'content-playback');
+  this.player.ads.endLinearAdMode();
+  assert.strictEqual(spy.callCount, 0, 'adend should not have fired');
+  this.player.ads.startLinearAdMode();
+  assert.strictEqual(this.player.ads.state, 'ad-playback');
+  this.player.ads.endLinearAdMode();
+  assert.strictEqual(spy.callCount, 1, 'adend should have fired');
+  this.player.trigger('playing');
+  assert.strictEqual(this.player.ads.state, 'content-playback');
+  this.player.ads.startLinearAdMode();
+  assert.strictEqual(this.player.ads.state, 'ad-playback');
+  this.player.trigger('adserror');
+  assert.strictEqual(spy.callCount, 2, 'adend should have fired');
+});
+
+QUnit.test('skipLinearAdMode in ad-playback does not trigger adskip', function(assert) {
+  assert.expect(10);
+  var spy = sinon.spy();
+  this.player.on('adskip', spy);
+  assert.strictEqual(this.player.ads.state, 'content-set');
+  this.player.trigger('adsready');
+  assert.strictEqual(this.player.ads.state, 'ads-ready');
+  this.player.tech.trigger('play');
+  this.player.ads.startLinearAdMode();
+  assert.strictEqual(this.player.ads.state, 'ad-playback');
+  this.player.ads.skipLinearAdMode();
+  assert.strictEqual(this.player.ads.state, 'ad-playback');
+  assert.strictEqual(spy.callCount, 0, 'adskip event should not trigger when skipLinearAdMode called in ad-playback state');
+  this.player.ads.endLinearAdMode();
+  assert.strictEqual(this.player.ads.state, 'content-resuming');
+  assert.strictEqual(this.player.ads.triggerevent, 'adend', 'The reason for content-resuming should have been adend');
+  this.player.trigger('playing');
+  assert.strictEqual(this.player.ads.state, 'content-playback');
+  assert.strictEqual(this.contentPlaybackSpy.callCount, 1, 'A content-playback event should have triggered');
+  assert.strictEqual(this.contentPlaybackReason(), 'playing', 'The reason for content-playback should have been playing');
```

Comment:
It's kind of hard to keep track of the flow in these last couple tests without a bit of whitespace.